Query block setup with block patterns integration

ntsekouras opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
Resolves: https://github.com/WordPress/gutenberg/issues/27031
Related: https://github.com/WordPress/gutenberg/issues/27575

This PR will enhance the initialisation of `Query` block, by setting up some `attributes` and selecting a starting `layout` either by using `block variations` or `block patterns` (some blocks to insert in the `Query Loop`).

For now we can select some basic attributes, which for now is the most important one `inherit` global query, and set accordingly the `post type` of the Query block.

## Block patterns integration
Since there are plans to integrate block patterns in different contexts (examples are setup context like the this PR, or transform context in Block Switcher later) we need to make sure it's clear for theme/block pattern authors how to make their block patterns available in each context/view.

### The proposed API

There are many similarities here to block variations, so I introduced in block patterns API a new property called `scope`.
This property will be an object with the available scopes as keys, like `inserter, block(for Placeholder)` etc..

`inserter` will be `boolean` and the `other scopes` `arrays` with the `block names` as values. This way patterns could be filtered in a rather generic way, with the new selector `__experimentalGetScopedBlockPatterns`.

example in a pattern:
```
'scope'      => array(
	'inserter' => false,
	'block'    => array( 'core/query' ),
),
```

If a block pattern has not set a `scope`, it will be handled as having `inserter` set to `true` for backwards compatibility.

I understand that patterns can have many blocks, but this will make it explicit where to show what and it's going to be more performant.

### Let's look at this with an example.
We have a pattern that includes blocks: `Columns, Image, Buttons, Quote , Paragraph`.
Without an explicit API, we will have to parse patterns to try to find matching blocks (performance).
The above pattern even though it might contain 5 blocks, it might be focused on the `Columns and/or Image` and we don't want to show it in the `Quote` block's related actions. This could be the exact opposite of course by focusing only in `Quote` for something like testimonials.

I'm not sure though how we will handle transforms eventually as it's a different operation than an initialization.

## Showcase

### Before
![queryBefore](https://user-images.githubusercontent.com/16275880/107400566-95ae3880-6b0a-11eb-8415-9d7ebc704289.gif)


### After
![this](https://user-images.githubusercontent.com/1204802/108528917-39020900-72d4-11eb-80a3-baaf0f8cec06.gif)


## Notes
1. Preview differs between page and FSE context like `index` template. This should be handled with examples probably like the `Post Title`.


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

ntsekouras committed:
first stab

mtias commented:
@jameskoster we might need to prioritize this one https://github.com/WordPress/gutenberg/issues/28735 

I think what we have in this PR is a great starting point, but I was thinking it'd make sense to have "Patterns + Blank" and when you pick Blank you see the "variations".

mtias commented:
This is an excellent start, thanks for integrating patterns!

mtias commented on packages/block-library/src/query/index.js:
```diff
@@ -22,6 +22,9 @@ export const settings = {
 	save,
 	variations,
+	isMatchingBlockPattern: ( pattern ) => {
+		return pattern.categories?.includes( 'core/query' );
```
Is this something we should make implicit? cc @youknowriad 

In the original issue I suggested `patterns: [ ...patternCategories ]`

ntsekouras replied:
By having a `matching` function will help us fine-tune the behaviour for different use cases. One thing that I'm almost certain will be needed for different contexts (ex. transforms with patterns) will be to also pass the block's attributes.

In addition, if we proceed with this approach and usage of patterns for initializing blocks, we should probably make a convention about the pattern's category name or even consider to augment the block patterns API to have some more properties. Maybe include the block's name somehow? 🤔 

It must be clear for theme/block pattern authors how to make their block patterns shown in this view (block setup)

gziolo replied:
For blocks, we have this `parent` field that allows the author of the block to tell where this block can be inserted. If the idea is similar here, we should try the same approach first. For patterns, it might `context` or `target` or something like that. It all depends on how we want to treat this type of pattern that might not make much sense outside of the given block.

> One thing that I'm almost certain will be needed for different contexts (ex. transforms with patterns) will be to also pass the block's attributes.

We can always add such a function in the future, for now, it feels like a simple match between the pattern and block should be enough.

One idea to consider in the future, similar to how it works with block variations. We could always register patterns together with a block but with PHP:

```
register_block_type_from_metadata( __DIR__, array(
    'patterns' => array(
        $my_pattern, // and so on
    ),
) );

ntsekouras replied:
>We can always add such a function in the future, for now, it feels like a simple match between the pattern and block should be enough.

I agree 100% :) 

>For blocks, we have this parent field that allows the author of the block to tell where this block can be inserted. If the idea is similar here, we should try the same approach first. For patterns, it might context or target or something like that.

For now, without having it explored in depth yet, the augmentation of `Patterns` API with a new prop makes more sense to me too. One thing that I have in mind is that we will have block patterns that will be used only on block setup, others on transforms etc..

I mention that because there are discussions about transforming a single block to a block pattern which might contain other blocks as well. I think the result would be to replace selected block in pattern and keep the others with their demo content.

This makes it similar with block variations about their `scope` attribute too. I think we could create a new property in `patterns` to include `block name` and such context. An example would be:

```
[ 'core/query-setup', 'core/buttons-setup', 'core/buttons-transforms'
```

I'll explore this a lot better after the designs.

mtias commented on lib/load.php:
```diff
@@ -128,3 +128,120 @@ function gutenberg_is_experiment_enabled( $name ) {
+// Demo query patterns.
+register_block_pattern_category(
+	'core/query',
+	array( 'label' => __( 'Core Query block', 'gutenberg' ) )
```
I think the patterns should be called "Post List" or something to that effect

ntsekouras replied:
These are just demo patterns that will not be included in the final code.

mtias commented:
@tellyworth this one is good to keep an eye on for the pattern directory integration into blocks.

mtias commented on lib/load.php:
```diff
@@ -128,3 +128,120 @@ function gutenberg_is_experiment_enabled( $name ) {
+	'core/query',
+	array( 'label' => __( 'Core Query block', 'gutenberg' ) )
+);
+register_block_pattern(
```
@kjellr can we take a look at the starting ones here? We can always tweak them at any point

ntsekouras replied:
Should we add patterns in GB or in TT1-blocks first?

kjellr replied:
I think bundling some generally-useful patterns here makes sense first (a list, a grid, some variations with and without post meta and images). I'll take some time tomorrow to look through the ones you've bundled here so far. 👍

We'll probably want to follow up this PR with some TT1-Blocks-specific patterns for that theme itself. cc @melchoyce so you can start thinking about those.

ntsekouras replied:
The way it works now for Query is to have block patterns that will be used **inside Query Loop**. What this means is that we should only include patterns for a **single post** ([check existing patterns example](https://github.com/WordPress/gutenberg/pull/28891/files#diff-979d356a31350e9429b6faa5be6daaa5b10785838520396f552717a1b270c67bR143)).

If we want to support this for list/grid etc, I'll have to check it out to make it work and see how the Previews will behave :)

kjellr replied:
Ah ok I see. The Variations show multiple posts at one time, so I figured the Patterns would work the same way: 

![Screen Shot 2021-02-11 at 9 05 41 AM](https://user-images.githubusercontent.com/1202812/107646807-630a5a00-6c48-11eb-8b86-0e3d13f54e9c.png)

This is actually one of those use cases where as a user I'm not really sure why there _are_ patterns and variations... they both do basically the same thing. For reference, here are the bundled patterns today: 

![Screen Shot 2021-02-11 at 9 14 38 AM](https://user-images.githubusercontent.com/1202812/107648089-e7111180-6c49-11eb-90a3-b736c8356cc6.png)

I was going to suggest we revise these bundled patterns to be simple, like: 

- Featured Image, Title, Excerpt stacked
- Featured Image, Title, Date stacked
- Title, Excerpt stacked
- Title, Date stacked
- (some variation of those in a grid if that's possible)

... but then I realized that half of those already exist as "Variations". If they're already covered there I'm not totally sure it makes sense to also have core provide patterns here. Maybe it only makes sense for the theme to provide its own opinionated patterns, and leave core's bundled "patterns" as "variations" instead? 

I'm curious to get other folks' thoughts on that.

ntsekouras replied:
>For reference, here are the bundled patterns today:

These are just copy/paste demo patterns from me. There are no patterns for Query yet anywhere.

The difference though even with these patterns is that they contain other blocks as well for better design, like `columns`.  Existing block variations contain just some `Post Blocks` listed one after the other.

kjellr replied:
Yeah, I see the benefit here. Here's a pass at some query block patterns we could consider bundling with Core to start with — they're a little more varied than the ones in the PR currently, and are just a little more opinionated/customized than the block variations: 

<img width="759" alt="01-full-width-image" src="https://user-images.githubusercontent.com/1202812/107815879-f6738600-6d41-11eb-8623-7cee10d20387.png">

<img width="1131" alt="02-image-on-left" src="https://user-images.githubusercontent.com/1202812/107815910-0723fc00-6d42-11eb-81a5-0b68b17142be.png">

<img width="578" alt="03-with-date" src="https://user-images.githubusercontent.com/1202812/107815882-f70c1c80-6d41-11eb-9b4a-6c14aaeda472.png">

<img width="549" alt="04-list" src="https://user-images.githubusercontent.com/1202812/107815884-f70c1c80-6d41-11eb-9736-67866c6e2800.png">

cc @melchoyce + @jasmussen for some thoughts on those. 

([Markup is available in this gist](https://gist.github.com/kjellr/a43502a5d4be823626dab250d3cd6865))

jasmussen replied:
This seems like a good start that doesn't preclude tweaks in the future. I think it's a very hard task because the patterns have to be super minimalist so as to work in any theme. I wonder if we should start even simpler, with just 3 starter patterns of varying levels of features. Such as:

- small (last one, has featured image and title only)
- medium (a variant of the 2nd one, image on the left, title, read more, perhaps no timestamp)
- big (a variant of the first one, title, wide image, timestamp, read more)

Also, we should provide some theme to these, so it doesn't read "Sample Post" and use lorem ipsum. I'll be happy to defer to you on the theme, but just throwing it out there, we reference Julio Cortázar in the Quote block description. Perhaps we could [cite some of his work](https://en.wikipedia.org/wiki/Julio_Cort%C3%A1zar#Bibliography)? Or does it have to be public domain?

kjellr replied:
I like the small/medium/large idea. 👍 

In terms of the content, that's just what loads in from my test site. 😄 Since these are all dynamic blocks I don't think we can specify something else, and I think that's fine. 

Here's the revised set: 

Large: 
<img width="657" alt="Screen Shot 2021-02-15 at 9 55 22 AM" src="https://user-images.githubusercontent.com/1202812/107961753-5487af80-6f74-11eb-886e-9cb84f330031.png">

Medium:
<img width="682" alt="Screen Shot 2021-02-15 at 9 55 31 AM" src="https://user-images.githubusercontent.com/1202812/107961775-5782a000-6f74-11eb-9dc2-61815b06f104.png">

Small:
<img width="668" alt="Screen Shot 2021-02-15 at 9 55 39 AM" src="https://user-images.githubusercontent.com/1202812/107961784-5baebd80-6f74-11eb-8bb7-cbad36c9f45b.png">

([Updated markup](https://gist.github.com/kjellr/63d2b7acd7f4e50c7d8f17d5a04b068b))

jasmussen replied:
Oh of course, I was thinking of patterns.

There's still something to think about entirely separately, for when you load the site editor on a site without content. But yes, for another time.

Very nice!

ntsekouras replied:
@kjellr I updated the patterns with a small change in `small` due to very bad spacing of `image placeholder`. I made the columns `25-75%`. We can revisit how an image placeholder is shown in really small spaces in a separate issue/PR. Is this okay?

mtias commented on packages/block-library/src/query/edit/block-setup/index.js:
```diff
@@ -0,0 +1,101 @@
+/**
+ * Internal dependencies
+ */
+import LayoutSetupStep from './layout-step';
```
It seems fine to start with the query block, but let's write the component with an eye towards making it part of the generic `<Placeholder>` whenever a block has registered patterns as per #28735.

ntsekouras replied:
I had this a bit in mind with `BlockSetup`. If we can consolidate the use cases of more blocks I'll look into it better. It's mostly a prototype now.

ntsekouras commented:
>I think what we have in this PR is a great starting point, but I was thinking it'd make sense to have "Patterns + Blank" and when you pick Blank you see the "variations".

We could use some design about this and also have in mind how it could be reused by other blocks.

jasmussen commented:
This is what I see:

![this](https://user-images.githubusercontent.com/1204802/107625894-49bfd880-6c5d-11eb-875a-dd0932c8b184.gif)

Powerful! It's also a big placeholder.

I'm going to take a quick stab at both a near term and longer term visual for this. Hopefully the near term mockup I can help push some CSS for directly. I'll be back.

jasmussen commented:
Here's a quick exercise. Based on your feedback I will adjust, and when we're happy, I can help push some CSS and we can get a V1 out.

**Initial setup**

<img width="1440" alt="Query step 1" src="https://user-images.githubusercontent.com/1204802/107631250-f487c500-6c64-11eb-8b40-16dcd4d46e12.png">

- A basic title, 2 buttons, and the "Inherit" toggle.
- The "Inherit" toggle, when untoggled, expands just as it does today. It shows conditional help text when untoggled, but not in its toggled state. 

**"Start empty" state**

You click "Start empty" in the first step, and you get this screen which you may recognize:

<img width="1440" alt="Query step 2a" src="https://user-images.githubusercontent.com/1204802/107631705-bccd4d00-6c65-11eb-8e2f-617d17389543.png">

**"Start with a pattern" state**

If you had clicked the primary "Start with a pattern" button, we show patterns right in the placeholder:

<img width="1440" alt="Query step 2b" src="https://user-images.githubusercontent.com/1204802/107631824-e8e8ce00-6c65-11eb-92ff-ec9d6da1f2ff.png">

The argument for a carousel is that it literally makes them bigger, literally bringing them closer to you. Instead of feeling like you pick a pattern, by being a carouse it feels like you've already chosen a pattern, now you're just paging through options before you confirm. 

Click the arrow keys to page between the available patterns.

<img width="1440" alt="Query step 2b 2" src="https://user-images.githubusercontent.com/1204802/107631875-fef68e80-6c65-11eb-92b5-c0fd9a219fdd.png">

Or click the pattern itself to insert it.

**Questions**

- Is this flow roughly a good way to start?
- Do we want a "Back" button in both the variation and pattern screens?

I think there's a lot more we can do to make this flow more intuitive. For example I'd love to potentially _start_ with the pattern browser directly — it helps the placeholder pass the "squint test". But for a V1, and per @mtias [thoughts on the flow](https://github.com/WordPress/gutenberg/pull/28891#issuecomment-776620664), keeping it two-step might be a good starting point?

ntsekouras commented:
>Do we want a "Back" button in both the variation and pattern screens?

Yes because the main thing here is the `block setup`, which can set up block attributes as well.

Also we need to have in mind that this could be used eventually in other blocks as well, like `columns`.

I'm a bit concerned about the `carousel` approach for a couple of reasons:
1. feels strange with the rest of the block interactions and current display of block patterns in other places (not necessarily a bad thing though 🤔 )
2. a user has to browse all the patterns one by one and will have to remember which one they liked etc.. - harder for visual comparison

mtias commented:
I think the carousel is something we need to discuss further on #28735 etc.

jasmussen commented:
Great thoughts. Here's a quick iteration based on feedback. It all still starts here:

<img width="1440" alt="Query step 1" src="https://user-images.githubusercontent.com/1204802/107644359-9fa17a00-6c77-11eb-925d-55a12ecc662c.png">

The "Start empty" state is the same, but now has a back button:

<img width="1440" alt="Query step 2a" src="https://user-images.githubusercontent.com/1204802/107644404-acbe6900-6c77-11eb-95ff-93829fd29ca6.png">

The "Patterns" state has a back button, and now shows 3 patterns side by side, click one to use it:

<img width="1440" alt="Query step 2b" src="https://user-images.githubusercontent.com/1204802/107644442-ba73ee80-6c77-11eb-958d-3da67df554a1.png">

I know that this branch features a scrollbar to afford more patterns. But I have a feeling that if we start with about 6 patterns at most, we can probably avoid it, give or take some responsive considerations. But I think this is best tackled directly in implementation.

Depending on additional feedback, I imagine next steps could be that I start helping out with some CSS here?

mtias commented:
The "start with a pattern" and "start empty" feels like a redundant step. I think I'd prefer to just show patterns and have blank be a choice among them.

ntsekouras commented:
Thanks for exploring the design for this 💯 

Just keep in mind that we will definitely need helper information about the specific `inherit` attribute and that there might be more controls for attributes set up. In this case we also have the conditionally shown `post type`.

If we keep patterns/attributes in one step, we should show the attributes always then. Also the attributes seem to me more important than layout, so it makes more sense to me to be on the top at such an approach.

gziolo commented on packages/block-library/src/query/edit/query-block-setup.js:
```diff
@@ -0,0 +1,109 @@
+			);
+		}
+	};
+	const blockSetupProps = {
```
In React, passing down complex objects like that usually doesn't scale well because of re-rendering. You can achieve a similar goal using a composition of smaller components and hooks that abstract common functionalities.

ntsekouras replied:
Yes! I'm waiting for the design and probably consolidate other blocks' use cases to refactor/change the code. We will definitely have more discussions about the implementation then :)

jasmussen commented:
Here's an iteration based on this feedback:

> The "start with a pattern" and "start empty" feels a bit redundant as a step. I think I'd prefer to just show patterns and have blank be a choice among them? 

<img width="1440" alt="Query_ Initial State" src="https://user-images.githubusercontent.com/1204802/107760864-bd7be700-6d2a-11eb-9a7d-dd410f40aa27.png">

Start empty still takes you here:

<img width="1440" alt="Query_ Start Empty" src="https://user-images.githubusercontent.com/1204802/107760874-c1a80480-6d2a-11eb-8230-6cbcbb7c0f1c.png">

> We'd need to handle narrow spaces, of course. (Maybe in those cases collapse into a button and show the inserter popup with patterns?)

Yep we have `is-small`,`is-medium` and `is-large` CSS classes we can use to show hide items here.

jasmussen commented:
After conversing with Nik, here are some tweaks. He noted that we need the help text for the "Inherit URL query" toggle, which currently has a rather verbose help text:

<img width="1440" alt="query with patterns" src="https://user-images.githubusercontent.com/1204802/107767913-f9687980-6d35-11eb-9345-ff64d4e3f9cb.png">

I can appreciate that, but I think there's an opportunity for a much more contextual and brief, but still descriptive help text. Such as:

<img width="296" alt="toggle on" src="https://user-images.githubusercontent.com/1204802/107768069-2ae14500-6d36-11eb-8093-33147653f819.png">

<img width="433" alt="toggle off" src="https://user-images.githubusercontent.com/1204802/107768072-2cab0880-6d36-11eb-8c58-08eeb0fea241.png">

I may be missing some nuance in the above verbiage, so feel free to suggest. But having the helptext be conditional to the toggle state, that alone will help make it less overwhelming.

---

Nik noted that there are cases where no patterns exist, or they weren't loaded. In that case, could we go directly to the variations screen?

<img width="1440" alt="start empty b" src="https://user-images.githubusercontent.com/1204802/107768212-67ad3c00-6d36-11eb-92c8-239eddea7db7.png">

The alternative becomes rather bare:

<img width="1440" alt="query without patterns" src="https://user-images.githubusercontent.com/1204802/107768302-83b0dd80-6d36-11eb-8bed-c5ae4aa1b1ca.png">

--- 

Nik also noted that some of the "back button" changes to the placeholder are non trivial to implement. Specifically referring to this screen you get to if you press "Start empty" on the first screen:

<img width="1440" alt="start empty a" src="https://user-images.githubusercontent.com/1204802/107768433-b3f87c00-6d36-11eb-9fe6-633516dab2bd.png">

An alternative has a non-integrated back button:

<img width="1440" alt="alternate start empty" src="https://user-images.githubusercontent.com/1204802/107768543-eb672880-6d36-11eb-9910-162148997f8b.png">

draganescu commented:
cc @shaunandrews @tellthemachines 
Does this kind of block setup also work for navigation? It would allow us to circumvent a lot of the problems with building complex menus, by adding such a step for selecting, say, header menu patterns or sidebar menu patterns.

jasmussen commented:
> Does this kind of block setup also work for navigation? I

To an extent, we have it with this:

<img width="657" alt="Screenshot 2021-02-12 at 14 16 45" src="https://user-images.githubusercontent.com/1204802/107772771-fa50d980-6d3c-11eb-8e31-359feea5651e.png">

The tricky challenge is that we can't use a giant white sheet for the navigation block, because it's almost never that big itself. Or to put it differently, the setup state should ideally never be bigger than the configured state.

I think there's an opportunity to revisit navigation with patterns directly inline, which would potentially be able to resize the setup state based on the size of the pattern you pick. @jameskoster just mocked up such a pattern carousel for the [Query block](https://github.com/WordPress/gutenberg/issues/28735#issuecomment-778179539), but presumably it could scale down to the navigation block also.

jameskoster commented:
> presumably it could scale down to the navigation block also

I think it could work! But I also think we need to do some dedicated exploration around how the patterns paradigm translates to navigation in general.

ntsekouras committed:
first iterations from design

Keyboard navigation, style changes, refactor LayoutStep, create `usePostTypes` util hook

ntsekouras committed:
remove duplicate imports from rebase

ntsekouras committed:
design iteration part 2 + keyboard navigation

ntsekouras committed:
Matching block patterns API v1

paaljoachim commented:
@ntsekouras Can you rebase the PR? I am trying to create a special Gutenberg plugin so that I can test it out. By going to the Checks (tab) -> Finding a mention of download Gutenberg in the left sidebar and then downloading it. 
It makes it easier for anyone to test out. Similar to was done here: https://github.com/WordPress/gutenberg/pull/28637#issuecomment-780868771

ntsekouras commented:
Hey @paaljoachim - I'm not sure how a rebase would help, although I have rebased some hours ago. You can still get a link from the build in the latest commit checks here: https://github.com/WordPress/gutenberg/actions/runs/578622288. That is what you meant, no?

gwwar reviewed (commented):
Really cool exploration @ntsekouras! ✨ This felt pretty intuitive for me. The one downside was wanting to go back to the pattern picker after I had selected something. It's probably out of scope for this exploration, but the carousel concepts, or having an easy way to toggle between patterns would be lovely. Right now I had to fumble through deleting the entire block then re-adding it to get back to the do-it-for-me wizard.

I also noticed that the patterns were also being registered in the post editor, so there might be some tidying needed to avoid that later if this wasn't intended.
<img width="892" alt="Screen Shot 2021-02-18 at 11 19 15 AM" src="https://user-images.githubusercontent.com/1270189/108412554-c7945e80-71de-11eb-9b49-82cb560f7bca.png">

Leaving some testing setup notes for other people who'd like to try this:

### Testing Instructions

#### Run code locally
- Run the branch on a local wp instance. This can be done by using a wp-env setup noted in [ getting-started ](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/getting-started.md#local-environment). 
- Alternatively, if you have a different local setup. Each PR also generates a plugin zip, which can be found by clicking the following:

| Plugin Zip Job | Download Artifact|
| ------------- | ------------- |
| <img width="736" alt="Screen Shot 2021-02-18 at 11 35 55 AM" src="https://user-images.githubusercontent.com/1270189/108411811-ee9e6080-71dd-11eb-808a-432ebb6179ae.png"> | <img width="1046" alt="Screen Shot 2021-02-18 at 11 33 47 AM" src="https://user-images.githubusercontent.com/1270189/108411840-f65e0500-71dd-11eb-8cb6-55003a459b58.png"> |

- Don't forget to check that the Gutenberg plugin is enabled

#### Enable the Site Editor

- The Query Block is used in the Site Editor
- Enable a theme that supports the site editor, such as TT1Blocks at `wp-admin/themes.php`
- With said theme activated, we should see: 

| Site Editor in Sidebar | 
| ------------- | 
| <img width="168" alt="Screen Shot 2021-02-18 at 11 41 59 AM" src="https://user-images.githubusercontent.com/1270189/108412211-6e2c2f80-71de-11eb-82a5-984d4659ca34.png"> |

- Click on the Site Editor
- Your index page will likely already be using the Query block, try deleting it and adding a new one.

https://user-images.githubusercontent.com/1270189/108410808-b4808f00-71dc-11eb-8857-38f8c2b8f0cc.mp4

ntsekouras commented:
Thanks for testing @gwwar!

>The one downside was wanting to go back to the pattern picker after I had selected something. It's probably out of scope for this exploration, but the carousel concepts, or having an easy way to toggle between patterns would be lovely

Yeah that will be handled later with :https://github.com/WordPress/gutenberg/issues/27575 and https://github.com/WordPress/gutenberg/issues/28735.

>I also noticed that the patterns were also being registered in the post editor, so there might be some tidying needed to avoid that later if this wasn't intended.

Yes again. I'm working on the API now: https://github.com/WordPress/gutenberg/pull/28891/commits/fb52a7272068d369dd92829026b2d435b9a49ea2 and this needs to handle the hiding in Inserter.

ntsekouras committed:
change API

jasmussen commented:
Really nice work. This is what I see:

<img width="1298" alt="Screenshot 2021-02-19 at 10 01 06" src="https://user-images.githubusercontent.com/1204802/108482191-785f3400-7299-11eb-8bfc-f2e154eb6bd4.png">

<img width="931" alt="Screenshot 2021-02-19 at 10 02 16" src="https://user-images.githubusercontent.com/1204802/108482278-92007b80-7299-11eb-8225-2687935f046b.png">

I have a bit on my plate today, otherwise I would've helped push some fixes. But from glancing at this it looks good, except for two things:

- we need to handle overflow and wrapping better. I'll circle back to that one.
- To match [this mockup](https://user-images.githubusercontent.com/1204802/107767913-f9687980-6d35-11eb-9345-ff64d4e3f9cb.png), the titles should all be on a row, and there should be a light gray border around the unhovered pattern
- I see five patterns, but Kjell started with 3 in https://github.com/WordPress/gutenberg/pull/28891#discussion_r576250473 — I think it'd be good to start with fewer and then add more only when we are confident. For those three, the titles could be Small, Medium and Large.
- The "Start Empty" should be the same size as patterns, but square.
- There are some margin padding issues.

**On the overflow**. I think this is a problem we probably shouldn't spend too much time on, as we might want to look at https://github.com/WordPress/gutenberg/issues/28735#issuecomment-780651352 some time in the future, and that would handle it. So I'm tempted to push a temporary fix, such as a horizontal scrollbar. However more importantly, I think we should start with only 3 patterns, which makes overflow handling less urgent. Then we can also refine the responsive styles there. 

I'd like to help you fix all of these by pushing CSS, but I can't do that until Monday. If you can wait, I'll come back to it.

ntsekouras committed:
move and change block patterns

ntsekouras commented:
>To match [this mockup](https://user-images.githubusercontent.com/1204802/107767913-f9687980-6d35-11eb-9345-ff64d4e3f9cb.png), the titles should all be on a row 

>The "Start Empty" should be the same size as patterns, but square.

I'm a bit concerned how to make the `previews` the same height as the content might be different in each of them and we are using `AutoHeightBlockPreview`. Also a previous implementation for fixed heights has been removed completely here: https://github.com/WordPress/gutenberg/pull/21096

In @youknowriad 's words:
>In this PR, I'm removing the old implementation entirely as it's more complex, more cumbersome without a big additional value.

Any suggestions Riad since you have worked on preview stuff?

jasmussen commented:
> I'm a bit concerned how to make the previews the same height as the content might be different in each of them and we are using AutoHeightBlockPreview.

Oh I was thinking of a CSS only change. I think I can get something to work here, I'll try and push something.

jasmussen committed:
Polish

jasmussen commented:
Okay I pushed some polish and got it to this point:

![this](https://user-images.githubusercontent.com/1204802/108528917-39020900-72d4-11eb-80a3-baaf0f8cec06.gif)

It's not quite what I want, and I'm not done here. But it's a start at aligning the things. I expect to push more things Monday (responsive, overflow, those double borders).

mtias commented:
A few things (don't have to happen now, but just to continue the design work):

- I think the "inherit from query" should be part of setup that handles things like "how many posts to show", "from which post type", etc. If we are adding the query block in "the content" it should probably never show "inherit query".
- The "start empty" should have more of an empty look and should be placed at the end?
- Given the complexity of everything going on and how much it requires extra space, perhaps it's worth having all this happen in a modal?

ntsekouras committed:
__experimentalGetScopedBlockPatterns doc and tests

jasmussen commented:
> Given the complexity of everything going on and how much it requires extra space, perhaps it's worth having all this happen in a modal?

We will probably need a modal at some point anyway, for the narrow viewports or contexts.

ntsekouras committed:
jsdoc usePostTypes

ntsekouras committed:
show start empty at the end

ntsekouras commented:
>The "start empty" should have more of an empty look and should be placed at the end?

I placed it at the end.

jasmussen committed:
Improve thumbnails.

jasmussen commented:
I pushed a fix to put the titles outside of the box:

<img width="892" alt="Screenshot 2021-02-22 at 14 37 10" src="https://user-images.githubusercontent.com/1204802/108715729-87551900-751b-11eb-874f-f4f8d7900ea1.png">

To address this one: 

> The "start empty" should have more of an empty look 

@ntsekouras can we do this?

<img width="899" alt="Screenshot 2021-02-22 at 14 36 39" src="https://user-images.githubusercontent.com/1204802/108715768-93d97180-751b-11eb-8eb9-dd3e77588561.png">

ntsekouras committed:
fix empty variation

jasmussen committed:
Polish the labels.

jasmussen commented:
Alright I pushed some polish:

![placeholder](https://user-images.githubusercontent.com/1204802/108719079-70b0c100-751f-11eb-8088-c3d9333bea1c.gif)

There are so many opportunities to polish this further — the icons aren't all that great, and the two-step layout maybe isn't ideal. But I think it's in a place no where it can be shipped, and be useful for a couple of releases, until we revisit this.

mtias commented:
Agreed on getting a baseline in place

mtias reviewed (approved):
Let's get this in, we have a lot of follow ups to do still but we should do that with more focused attention on separate issues

ockham commented:
cc/ @simison -- The new block patterns API that this PR introduces might make sense for migrating some Gutenboarding (and/or WP.com block patterns) functionality to it.
